Add a dynamic services target with a variant for the embedder and dylib each

chinmaygarde opened the pull request:
* The embedder depends on dynamic:embedder
* The dylib depends on dynamic:dylib
* The embedder calls OnLoad and OnUnload callbacks that setup and teardown the dylib environment
* The dylib contains a thin library that services OnLoad, OnUnload and OnInvoke so that vendors dont have to do it themselves (and potentially mess it up)
* The vendor only has to implement the `FlutterServicePerform` method that takes a fully scoped handle
* This patch is a WIP till we get a stable Flutter ABI. The stuff in mojo/public is *NOT* stable

chinmaygarde commented:
cc @abarth @jamesr @mpcomplete

mpcomplete commented on sky/services/dynamic/dynamic_service_dylib.cc:
```diff
@@ -0,0 +1,47 @@
+#include "sky/services/dynamic/dynamic_service_dylib.h"
+#include "mojo/public/cpp/environment/environment.h"
+
+static mojo::Environment* GDylibEnvironment;
```
g_dylib_environment

mpcomplete commented on sky/shell/platform/ios/sky_dynamic_service_loader.mm:
```diff
@@ -94,6 +91,8 @@ - (void)openIfNecessary {
     return;
   }
 
+  // Set 1: Find the dylib containing the service implementation
```
Step 1

mpcomplete replied:
End comments with a '.'

jamesr commented on sky/services/dynamic/dynamic_service_dylib.cc:
```diff
@@ -0,0 +1,47 @@
+
+#include "sky/services/dynamic/dynamic_service.h"
+
+#include <cassert>
```
you should probably use <assert.h>. the assert header defines a macro so the namespacing functionality that most <cfoo> headers provide is extra useless here.

jamesr commented on sky/services/dynamic/dynamic_service_dylib.cc:
```diff
@@ -0,0 +1,47 @@
+  // The service is always responsible for releasing the client handle. Create
+  // a scoped handle immediately.
+  mojo::MessagePipeHandle message_pipe_handle(client_handle);
+  mojo::ScopedMessagePipeHandle scoped_handle(message_pipe_handle);
```
do you know in this layer what type this handle is supposed to be (i.e. does it correspond to a particular mojom interface)? if so you probably want to go ahead to that type here to avoid confusing later

chinmaygarde replied:
No, `FlutterServiceInvoke` is generic to all services.

jamesr commented on sky/services/dynamic/dynamic_service_dylib.cc:
```diff
@@ -0,0 +1,47 @@
+  mojo::ScopedMessagePipeHandle scoped_handle(message_pipe_handle);
+
+  if (g_dylib_environment == nullptr) {
+    return;
```
should this be an assert? is it valid to try to Invoke before calling OnLoad?

jamesr commented on sky/shell/platform/ios/sky_dynamic_service_loader.mm:
```diff
@@ -156,18 +182,18 @@ - (SkyDynamicServiceHandler)serviceEntryPoint {
+  void* entry = dlsym(_libraryHandle, kFlutterServiceInvokeProcName);
 
   if (entry == NULL || dlerror() != NULL) {
     LOG(INFO) << "Could not find service entry point '"
```
I think it'd be cleaner to look up all entry points on initially loading and not call in to the DLL until you know that all of the symbols you might want to invoke are exported. you'll still run dll initializers inside the dll before checking the symbols, which you can't avoid (unless you patch the linker itself) but you can simplify error handling a lot by avoid having partially-initialized but invalid libraries loaded

chinmaygarde replied:
Yeah, had the same feeling. This file used to be so simple :( Will separate this out.

jamesr commented on sky/services/dynamic/dynamic_service.h:
```diff
@@ -0,0 +1,23 @@
+
+FLUTTER_C_API_START
+
+FLUTTER_EXPORT void FlutterServiceOnLoad(const MojoAsyncWaiter* waiter,
```
this isn't valid C. in C, you would have to spell this type name

const struct MojoAsyncWaiter* waiter

and write (void) instead of () in the Unload declaration, etc. if you want to verify that these things compile as C (and not just as C++ in an extern "C" declaration) you should probably have a .c file somewhere that #includes the headers you want to be C. or you could say that these APIs are only usable from C++ / Obj-C++, but that might be more annoying for consumers.

chinmaygarde replied:
> you should probably have a .c file somewhere that #includes the headers you want to be C

Good idea. Will do.

jamesr commented on sky/services/dynamic/BUILD.gn:
```diff
@@ -0,0 +1,35 @@
+
+  deps = [
+    "//mojo/public/c/system:system",
+    "//mojo/public/c/system:system",
```
this is doubly redundant.  you can just say "//mojo/public/system" as //foo/bar is defined to be the same thing as //foo/bar:bar and you have this line listed twice

jamesr commented on sky/services/dynamic/dynamic_service_dylib.cc:
```diff
@@ -0,0 +1,47 @@
+  mojo::String name(service_name);
+
+  if (name.size() == 0) {
+    return;
```
the service name comes from the flutter engine, right? if the engine passes bogus data in i'm not sure you want the dylib to have to try to deal with it, probably better to have all of the validation happen on the engine side

chinmaygarde replied:
You're right. We already validate this on the embedder side

jamesr commented on sky/services/dynamic/dynamic_service_dylib.h:
```diff
@@ -0,0 +1,19 @@
+FLUTTER_C_API_START
+
+void FlutterServicePerform(mojo::ScopedMessagePipeHandle client_handle,
+                           const mojo::String& service_name);
```
this is not C. does this symbol need to avoid mangling for some reason?

chinmaygarde replied:
No. I forgot to take these out when I split the library to be embedder and dylib specific. Removed.

jamesr commented:
Several parts of //mojo/public are binary stable, just not all of it.  It might be useful to sketch out which parts of this API are temporary to be replaced so we can check that the parts not marked out are in fact stable.

mpcomplete commented:
Maybe add a README that explains what parts of this are public API? It's not immediately clear to me without reading your commit msg.

chinmaygarde commented:
@jamesr: I have addressed all your feedback. Now, unless all service procs are found in the library, the system thunks nor the on load methods are invoked on the library. See the `DynamicServiceDefinition::Setup` and `Teardown` methods for the steps followed.

@mpcomplete: I have added a simple README. I will keep updating it as the steps change.

Another look please.

jamesr commented on sky/services/dynamic/BUILD.gn:
```diff
@@ -0,0 +1,44 @@
+  ]
+
+  deps = [
+    "//mojo/public/c/environment:environment",
```
redudant, kill everything after the :

jamesr commented on sky/services/dynamic/BUILD.gn:
```diff
@@ -0,0 +1,44 @@
+
+  deps = [
+    "//mojo/public/c/environment:environment",
+    "//mojo/public/c/system:system",
```
ditto here and further down

chinmaygarde replied:
"//mojo/public/cpp/environment:standalone" and "//mojo/public/platform/native:system" were required since they are not being added using the default rule. I have removed the rest of the redundancies in this file. Thanks.

jamesr commented on sky/services/dynamic/dynamic_service_definition.cc:
```diff
@@ -0,0 +1,178 @@
+}
+
+void DynamicServiceDefinition::Setup(const mojo::String& dylib_path) {
+  std::lock_guard<std::mutex> lock(lifecycle_mtx_);
```
what's this lock for? seems pretty scary to be holding a mutex during dylib load/unload depending on what the dylib does. what sort of synchronous calls can the dylib make into the engine with this lock held?

chinmaygarde replied:
If the service provider calls our dynamic service resolver multiple times from different threads. There may be a case were we install thunks and call the onLoad/Unload methods multiple times.

chinmaygarde replied:
> what sort of synchronous calls can the dylib make into the engine with this lock held?

Presumably, the dylib can request dynamic service resolution of more services. This is fine, as the lock is per dylib. There may be a deadlock in case the service requests resolution of itself in the OnLoad callback. But there is little to do to guard against that. This is extremely unlikely though as we write the default OnLoad method in `dynamic_service_dylib`. So the service vendor will have to do something really nasty (maybe resolving itself in an attr constructor or something) to cause an issue here

chinmaygarde commented:
Because of no-l.g.t.m love, I have amended this PR to include service versioning checks. In case of breaking changes to the Mojo or Flutter ABIs, we can refuse to load conflicting services. Just versioning stuff is in https://github.com/chinmaygarde/flutter_engine/commit/a504f10004dfcc3dd9a6ee968f610d49970079a1

abarth commented:
LGTM but jamesr is really your best reviewer.

chinmaygarde commented:
Rebased. Will address issues as they are brought up. Landing this to continue work on the sample.
